packages/blade/src/utils/useBreakpoint/useBreakpoint.ts

```diff
@@ -0,0 +1,149 @@
+import { useEffect, useState, useCallback, useMemo } from 'react';
+import type { Breakpoints } from '../../tokens/global';
+
+const deviceType = {
+  desktop: 'desktop',
+  mobile: 'mobile',
+} as const;
+
+type DeviceType = keyof typeof deviceType;
+type Breakpoint = keyof Breakpoints | undefined;
+
+type BreakpointAndDevice = {
+  matchedBreakpoint: Breakpoint;
+  matchedDeviceType: DeviceType;
+};
+
+const useBreakpoint = ({ breakpoints }: { breakpoints: Breakpoints }): BreakpointAndDevice => {
+  const supportsMatchMedia =
+    typeof document !== 'undefined' && typeof window?.matchMedia === 'function';
+
+  const breakpointsTokenAndQueryCollection = useMemo(
+    () =>
+      (supportsMatchMedia
+        ? Object.entries(breakpoints).map(([token, screenSize], index, breakpointsArray) => {
+            let mediaQuery = '';
+
+            if (token === 'max') {
+              mediaQuery = `screen and (min-width: ${screenSize}px)`;
+            } else if (breakpointsArray[index - 1]) {
+              mediaQuery = `screen and (min-width: ${
+                breakpointsArray[index - 1][1] + 1
+              }px) and (max-width: ${screenSize}px)`;
+            } else {
+              mediaQuery = `screen and (max-width: ${screenSize}px)`;
+            }
+
+            return { token, screenSize, mediaQuery };
+          })
+        : []) as {
```

Comment:
We have to do this because `Object.entries(breakpoints)` is inferring the type as `[string, number]`. Can it not retain ` "xs" | "s" | "m" | "l" | "xl" | "max"` instead of a generic string here?

// @anuraghazra any ideas?

Reply:
Yes we can do that have to write a stricter objectEntries() function to infer the types. 

Same as we do for objectKeys.
```ts
const objectKeys = <T extends Record<string, unknown>>(obj: T) =>
  (Object.keys(obj) as unknown) as (keyof T)[]
```

Reply:
Okay yeah this should work. 

```ts
const objectEntries = <T extends Record<string, any>>(obj: T) =>
  (Object.entries(obj) as unknown) as [keyof T, T[keyof T]][];
```